packages/flutter_markdown/lib/src/markdown.dart

```diff
@@ -0,0 +1,472 @@
+
+    if (_currentBlock.tag == 'pre')
+      top.add(_syntaxHighlighter.format(text.text));
+    else
+      top.add(text.text);
+  }
+
+  bool visitElementBefore(md.Element element) {
+    if (_isListTag(element.tag))
+      _listIndents.add(element.tag);
+
+    if (_isBlockTag(element.tag)) {
+      List<_Block> blockList;
+      if (_currentBlock == null)
+        blockList = _blocks;
+      else
+        blockList = _currentBlock.subBlocks;
+
+      _Block newBlock = new _Block(element.tag, element.attributes, _markdownStyle, new List<String>.from(_listIndents), blockList.length);
+      blockList.add(newBlock);
+    } else {
+      TextStyle style = _markdownStyle._styles[element.tag];
+      if (style == null)
+        style = new TextStyle();
+
+      List<dynamic> styleElement = <dynamic>[style];
+      _currentBlock.stack.add(styleElement);
+    }
+    return true;
+  }
+
+  void visitElementAfter(md.Element element) {
+    if (_isListTag(element.tag))
+      _listIndents.removeLast();
+
+    if (_isBlockTag(element.tag)) {
+      if (_currentBlock.stack.length > 0) {
+        _currentBlock.stack = _currentBlock.stack.first;
+        _currentBlock.open = false;
+      } else {
+        _currentBlock.stack = <dynamic>[''];
+      }
+    } else {
+      if (_currentBlock.stack.length > 1) {
+        List<dynamic> popped = _currentBlock.stack.last;
+        _currentBlock.stack.removeLast();
+
+        List<dynamic> top = _currentBlock.stack.last;
+        top.add(popped);
+      }
+    }
+  }
+
+  static const List<String> _kBlockTags = const <String>['p', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'li', 'blockquote', 'img', 'pre', 'ol', 'ul'];
+  static const List<String> _kListTags = const <String>['ul', 'ol'];
+
+  bool _isBlockTag(String tag) {
+    return _kBlockTags.contains(tag);
+  }
+
+  bool _isListTag(String tag) {
+    return _kListTags.contains(tag);
+  }
+
+  _Block get _currentBlock => _currentBlockInList(_blocks);
+
+  _Block _currentBlockInList(List<_Block> blocks) {
+    if (blocks.length == 0)
+      return null;
+
+    if (!blocks.last.open)
+      return null;
+
+    _Block childBlock = _currentBlockInList(blocks.last.subBlocks);
+    if (childBlock != null)
+      return childBlock;
+
+    return blocks.last;
+  }
+}
+
+class _Block {
+  _Block(this.tag, this.attributes, this.markdownStyle, this.listIndents, this.blockPosition) {
+    TextStyle style = markdownStyle._styles[tag];
+    if (style == null)
+      style = new TextStyle(color: Colors.red[500]);
+
+    stack = <dynamic>[<dynamic>[style]];
+    subBlocks = <_Block>[];
+  }
+
+  final String tag;
+  final Map<String, String> attributes;
+  final MarkdownStyle markdownStyle;
+  final List<String> listIndents;
+  final int blockPosition;
+
+  List<dynamic> stack;
+  List<_Block> subBlocks;
+
+  bool get open => _open;
+  void set open(bool open) {
+    _open = open;
+    if (!open && subBlocks.length > 0)
+      subBlocks.last.last = true;
```

Comment:
I'd probably rename the `last` property so that this line doesn't look so goofy.